utils/cloud_metadata.py

```diff
@@ -218,6 +238,31 @@ def get_metadata(agentConfig):
                 log.debug("Collecting EC2 Metadata failed %s", str(e))
                 pass
 
+        # Only check if we don't know yet or if it is known to be OpenStack
+        if EC2.is_openstack is None or EC2.is_openstack is True:
+            log.info(u"Attempting to get OpenStack meta_data.json")
+            openstack_metadata_url = EC2.EC2_METADATA_HOST + "/openstack/latest/meta_data.json"
+            try:
+                r = requests.get(openstack_metadata_url, timeout=EC2.TIMEOUT)
+                r.raise_for_status() # Fail on 404 etc
+
+                EC2.is_openstack = True
+                openstack_metadata = r.json()
+                # Set the "also known as" metadata similar to AWS EC2
+                EC2.metadata['host_aliases'] = [
+                    openstack_metadata['uuid'],
+                    openstack_metadata['name'],
+                ]
+
+                # the OpenStack instance-id can be recycled but the uuid will always be unique
+                if agentConfig['openstack_use_uuid']:
+                    log.info(u"Using the instance's uuid for instance-id")
+                    EC2.metadata["instance-id"] = openstack_metadata['uuid']
```

Comment:
I am setting it as a host alias in the lines above.

The reason I'm overriding it is because the default `i-####` value in openstack is *not* stable and is recycled by Nova. I did this as a flag so that it is backwards compatible though I'd hope that Nova would stop recycling the values. :(